candle-transformers/src/models/stable_diffusion/euler_ancestral_discrete.rs

```diff
@@ -0,0 +1,227 @@
+                if *n > step_ratio {
+                    Some(n - step_ratio)
+                } else {
+                    None
+                }
+            })
+            .map(|n| n - 1)
+            .collect(),
+            TimestepSpacing::Linspace => {
+                super::utils::linspace(0.0, (config.train_timesteps - 1) as f64, inference_steps)?
+                    .to_vec1::<f64>()?
+                    .iter()
+                    .map(|&f| f as usize)
+                    .rev()
+                    .collect()
+            }
+        };
+
+        let betas = match config.beta_schedule {
+            BetaSchedule::ScaledLinear => super::utils::linspace(
+                config.beta_start.sqrt(),
+                config.beta_end.sqrt(),
+                config.train_timesteps,
+            )?
+            .sqr()?,
+            BetaSchedule::Linear => {
+                super::utils::linspace(config.beta_start, config.beta_end, config.train_timesteps)?
+            }
+            BetaSchedule::SquaredcosCapV2 => betas_for_alpha_bar(config.train_timesteps, 0.999)?,
+        };
+        let betas = betas.to_vec1::<f64>()?;
+        let mut alphas_cumprod = Vec::with_capacity(betas.len());
+        for &beta in betas.iter() {
+            let alpha = 1.0 - beta;
+            alphas_cumprod.push(alpha * *alphas_cumprod.last().unwrap_or(&1f64))
+        }
+        let sigmas: Vec<f64> = alphas_cumprod
+            .iter()
+            .map(|&f| ((1. - f) / f).sqrt())
+            .collect();
+
+        let sigmas_xa = (0..sigmas.len()).map(|i| i as f64).collect();
+
+        let mut sigmas_int = interp(
+            &timesteps.iter().map(|&t| t as f64).collect(),
+            &sigmas_xa,
+            &sigmas,
+        );
+        sigmas_int.push(0.0);
+
+        // standard deviation of the inital noise distribution
+        // f64 does not implement Ord such that there is no `max`, so we need to use this workaround
+        let init_noise_sigma = *sigmas_int
+            .iter()
+            .chain(std::iter::once(&0.0))
+            .reduce(|a, b| if a > b { a } else { b })
+            .expect("init_noise_sigma could not be reduced from sigmas - this should never happen");
+
+        Ok(Self {
+            sigmas: sigmas_int,
+            timesteps,
+            init_noise_sigma,
+            config,
+        })
+    }
+
+    pub fn timesteps(&self) -> &[usize] {
+        self.timesteps.as_slice()
+    }
+
+    /// Ensures interchangeability with schedulers that need to scale the denoising model input
+    /// depending on the current timestep.
+    ///
+    /// Scales the denoising model input by `(sigma**2 + 1) ** 0.5` to match the K-LMS algorithm
+    pub fn scale_model_input(&self, sample: Tensor, timestep: usize) -> Result<Tensor> {
+        let step_index = self
+            .timesteps
+            .iter()
+            .position(|&t| t == timestep)
+            .ok_or_else(|| {
+                Error::Msg(format!(
```

Comment:
This one is a little bit tricky, `bail` macro return a `Result` , but `ok_or_else` expect `Error`.  But Let use `match` to do it instead of `ok_or_else`.